docs/03-code-internals/19-service-objects.md

```diff
@@ -0,0 +1,622 @@
+
+### `model`
+
+This specialized step helps to remove some boilerplate when dealing with models. By default, it will execute the method named `fetch_<name>`. In the above example, you can see we name our model `:user` and the corresponding method is named `fetch_user`.
+
+Here, you can fetch (or instantiate) a model as you see fit. If the step returns a falsy value, then the execution flow will stop here. If an `ActiveRecord` model is returned, it will call `#invalid?` on it to determine whether the model is valid. If not, the execution flow will stop.
+
+This step is also compatible with collections: if a collection is fetched but empty, the execution flow will stop.
+
+You can also provide an `ActiveRecord` relation. The step won’t load records but will determine if the relation will return any records. If there are none, the execution flow will stop.
+
+Sometimes, you need to fetch a model (or a collection of models), but it’s ok if it’s empty. For those cases, you can use the `optional: true` option allowing the execution flow to continue even if the model returns a falsy value.
+
+### `policy`
+
+This step will execute the method of the same name. You can put arbitrary code here, and the execution flow will stop if the return value is falsy.
+
+Usually, a policy is related to some state on one of the service models and/or to the current user (if any).
+
+### `params`
+
+This is one of the most powerful steps. Its main purpose is to validate the incoming data before feeding it to the models and to the service at a more global level. This is actually `ActiveModel` validations but applied to the incoming parameters.
+
+This step will run coercions and validations defined in the provided block. If the underlying contract isn’t valid (at least one validation failed), then the execution flow will stop.
+
+### `transaction`
+
+This step is a bit special, as it will wrap any other steps defined in its block inside a SQL transaction.
+
+### `options`
+
+This step is another special one, as it’s similar to a contract (without the validations part), but for options your service can take. This is useful if you need to change your service behavior depending on certain conditions. Also, that step can’t fail.
+
+### Steps arguments
+
+Each step is called with the service context. To access a value in it, just provide its key as a keyword argument.
+
+## The context object
+
+The only state a service maintain is its context object. This is where each step can put data to be used by other steps. Most of the time, you don’t need to access the context directly, as specialized steps will store the proper data for you. But sometimes it’s necessary. In those cases, it’s just a matter of using the context like you would with a hash:
+```ruby
+def first_step
+  context[:my_special_key] = "My special value"
+end
+
+def second_step(my_special_key:)
+  # do something with `my_special_key`
```

Comment:
We should make it clear how we assign value here.

Should it now be:

```ruby
def second_step(my_special_key:)
  my_special_key = "foo"
  
  # or
  
  context[:my_special_key] = "foo"
end
```

This is why I was telling you it's a little bit weird to not pass `def second_step(context:)` as it creates this ambiguous possibility.

Reply:
Not sure to follow, as this is just standard ruby. `my_special_key` is a variable local to the method scope here, so it won’t change anything globally to assign something to it inside the method body.  
I can expand a bit on the context being available globally, though. Or are you saying we shouldn’t make the context available globally, and if you want to access it, you have to ask for it too? (Just saw you wrote `context:` and not `context` :sweat_smile:)

Reply:
It’s adding boilerplate for virtually no gain, as we’re using the context everywhere. Having a nicer way to access some keys from it is better IMO, it allows taking advantage of the hash shorthand syntax too. It also communicates what are your step dependencies in a way at the definition level and if it can't match the key from the context it will raise (as it should as this is then a bug).